cognite/client/_api/sequences.py

```diff
@@ -350,38 +454,22 @@ def retrieve_dataframe(
         Returns:
              pandas.DataFrame
         """
-        pd = utils._auxiliary.local_import("pandas")
-        col, data = self._retrieve_with_columns(start, end, columns, external_id, id)
+        return self.retrieve(start, end, columns, external_id, id).to_pandas()
 
-        transposed_data = zip(
-            *[[d or math.nan for d in row["values"]] for row in data]
-        )  # make sure string columns don't have 'None'
-        data_dict = {k: list(v) for k, v in zip(col, transposed_data)}
-        return pd.DataFrame(index=[d["rowNumber"] for d in data], data=data_dict)
-
-    def _retrieve_with_columns(
-        self, start: int, end: int, columns: List[Union[int, str]] = None, external_id: str = None, id: int = None
-    ) -> Tuple[List, List[dict]]:
-        utils._auxiliary.assert_exactly_one_of_id_or_external_id(id, external_id)
-        post_obj = self._process_ids(id, external_id, wrap_ids=True)[0]
-        post_obj.update(self._process_columns(columns))
-        post_obj.update({"inclusiveFrom": start, "exclusiveTo": end})
-        return self._fetch_data(post_obj)
-
-    def _fetch_data(self, task):
-        seqdata = []
+    def _fetch_data(self, task, callback):
         task["limit"] = task.get("limit") or self._calculate_limit(task)
+        columns = []
         while True:
             items = self._post(url_path=self._DATA_PATH + "/list", json={"items": [task]}).json()["items"]
             data = items[0]["rows"]
-            columns = [c.get("externalId") or c["id"] for c in items[0]["columns"]]
-            seqdata.extend(data)
+            columns = columns or items[0]["columns"]
+            callback(data)
             if len(data) < task["limit"]:
                 break
             task["inclusiveFrom"] = data[-1]["rowNumber"] + 1
             if task["exclusiveTo"] and task["inclusiveFrom"] >= task["exclusiveTo"]:
                 break
-        return columns, seqdata
+        return columns
```

Comment:
I don't really like this callback pattern here, the method is called `fetch_data` but does not return the data which is a little confusing. It makes the code less readable. I would prefer if you return columns and seqdata as before and handle the resulting seqdata in the calling method instead of with the callback.

Reply:
needed for not keeping all data in memory, as discussed

Reply:
Could just make it into a generator instead?